### PR TITLE
NET60-13: Rewrite without MMA and upgrade to .NET 6

### PR DESCRIPTION
**Pull request by repl-dheeraj-v**
Test plan:
1. Verify that the existing unit tests are passing

image

Performance comparison:
I used the following code to measure the performance of both the assemblies. 
```
using Replicon.Cryptography.SCrypt;

namespace Program
{
    public class Program
    {
        public static void Main() 
        {
            IKeyDerivationFunction kdf = Factory.CreateKeyDerivationFunction();
            ulong[] memoryCost = { 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384 };
            uint r = 8, p = 1;
            for (int i = 0; i < memoryCost.Length; i++)
            {
                Stopwatch sw = Stopwatch.StartNew();
                kdf.DeriveKey(Encoding.ASCII.GetBytes("pleaseletmein"), Encoding.ASCII.GetBytes("SodiumChloride"), memoryCost[i], r, p, 64);
                sw.Stop();
                TimeSpan ts = sw.Elapsed;

                // Format and display the TimeSpan value.
                string elapsedTime = String.Format("{0:00}:{1:00}:{2:00}.{3:000}",
                    ts.Hours, ts.Minutes, ts.Seconds,
                    ts.Milliseconds);
                Console.WriteLine("N: " + memoryCost[i] + " r: " + r + " p: " + p + " RunTime:" + elapsedTime);
            }
        }
    }
}
```
a. Time taken by MMA implementation
image
b. Time taken by pure C# implementation using code from https://github.com/viniciuschiele/Scrypt
image 
c. Time taken by pure C# implementation when using https://www.nuget.org/packages/Bryllite.Cryptography.Scrypt/0.6.0 (drawback: only supports .NET Framework 4.8 and .NET Core 2 and above)
image
image

We have to ignore the first data point in both the tests because there is a delay by JIT compiler in the first execution.

**By running the above test for a number of times I can say that the pure C# implementation takes like** 
- **10x to 14x longer than the MMA implementation in case of https://github.com/viniciuschiele/Scrypt.**
- **approx 1.5x to 2x longer than the MMA implementation in case of https://www.nuget.org/packages/Bryllite.Cryptography.Scrypt/0.6.0**

### CODE REVIEW

File: Replicon.Cryptography.SCrypt/DefaultPasswordHash.cs [hunk 1]
```diff
@@ -121,13 +121,17 @@ private static SaltParseException InternalTryParseSalt(string salt, out byte[] s
...
         {
+#pragma warning disable CS8601 // Possible null reference assignment.
             var error = InternalTryParseSalt(salt, out saltBytes, out N, out r, out p, out hashLengthBytes);
+#pragma warning restore CS8601 // Possible null reference assignment.
             return error == null;
         }
 
         public void ParseSalt(string salt, out byte[] saltBytes, out ulong N, out uint r, out uint p, out uint hashLengthBytes)
         {
+#pragma warning disable CS8601 // Possible null reference assignment.
```
Comment (repl-mathieu-fenniak) [inline Replicon.Cryptography.SCrypt/DefaultPasswordHash.cs:132]: Ditto

File: Replicon.Cryptography.SCrypt/KeyDerivationFunction.cs [hunk 2]
```diff
@@ -0,0 +1,311 @@
+﻿using System;
+using System.Security.Cryptography;
+
+namespace Replicon.Cryptography.SCrypt
+{
+    //FIXME: Use https://github.com/viniciuschiele/Scrypt as a dependency after PR #11 is released.
```
Comment (repl-mathieu-fenniak) [inline Replicon.Cryptography.SCrypt/KeyDerivationFunction.cs:6]: Please add to this note that the below file is copied from that project, include the original author's name and a reference to the license under which we believe it is permitted to use this code.

**Review by repl-mathieu-fenniak — CHANGES_REQUESTED** [review]
As mentioned in our chat discussion, the testing needs to include:
- testing on multiple platforms (Windows, Mac, Linux)
- performance testing with the MMA implementation as a baseline

File: Replicon.Cryptography.SCrypt/DefaultPasswordHash.cs [hunk 3]
```diff
@@ -121,13 +121,17 @@ private static SaltParseException InternalTryParseSalt(string salt, out byte[] s
 
         public bool TryParseSalt(string salt, out byte[] saltBytes, out ulong N, out uint r, out uint p, out uint hashLengthBytes)
         {
+#pragma warning disable CS8601 // Possible null reference assignment.
```
[the remark this answers is not in the document]
Comment (repl-dheeraj-v) [reply Replicon.Cryptography.SCrypt/DefaultPasswordHash.cs:124]: Already throwing `SaltParseExcpetion`, so removing the #pragma directive

File: Replicon.Cryptography.SCrypt/DefaultPasswordHash.cs (hunk 1 above)
Comment (repl-dheeraj-v) [reply Replicon.Cryptography.SCrypt/DefaultPasswordHash.cs:132]: Already throwing `SaltParseExcpetion`, so removing the #pragma directive

File: Replicon.Cryptography.SCrypt/IPasswordHash.cs [hunk 4]
```diff
@@ -1,9 +1,7 @@
-﻿using System;
-
+﻿
 namespace Replicon.Cryptography.SCrypt
-{
-    /// <summary>Wrapper for the scrypt key-derivation function that provides helper functions for a common use-case
-    /// of scrypt as a password hashing algorithm.</summary>
+{/// <summary>Wrapper for the scrypt key-derivation function that provides helper functions for a common use-case
```
[the remark this answers is not in the document]
Comment (repl-dheeraj-v) [reply Replicon.Cryptography.SCrypt/IPasswordHash.cs:3]: Done

**Comment by repl-dheeraj-v** [comment]
> As mentioned in our chat discussion, the testing needs to include:
> 
> * testing on multiple platforms (Windows, Mac, Linux)
> * performance testing with the MMA implementation as a baseline

For testing on multiple platforms, can we do it on appveyor?

File: Replicon.Cryptography.SCrypt/KeyDerivationFunction.cs (hunk 2 above)
Comment (repl-dheeraj-v) [reply Replicon.Cryptography.SCrypt/KeyDerivationFunction.cs]: Done

**Review by repl-mathieu-fenniak — CHANGES_REQUESTED** [review]
Based upon these testing numbers, the performance of this implementation is 13x worse in the default scrypt configuration.  How can we move forward with such a dramatic performance degradation?  There is a significant risk with that change that either needs risk mitigation, or for an alternate higher speed implementation.

Regarding the MacOS & Linux testing; WSL is OK for Linux, and for MacOS reach out to a coworker with a Mac to perform dotnet tests.

**Comment by repl-dheeraj-v** [comment]
> Based upon these testing numbers, the performance of this implementation is 13x worse in the default scrypt configuration. How can we move forward with such a dramatic performance degradation? There is a significant risk with that change that either needs risk mitigation, or for an alternate higher speed implementation.
> 
> Regarding the MacOS & Linux testing; WSL is OK for Linux, and for MacOS reach out to a coworker with a Mac to perform dotnet tests.

I have used a new library which has improved the performance.
Build result on Mac:
Screenshot 2022-08-24 at 2 46 23 PM

**Review by repl-mathieu-fenniak — APPROVED** [review]

**Review by repl-mathieu-fenniak — APPROVED** [review]

**Pull request merged**